### PR TITLE
Fix `BackingStore` race and `Volatile` race

### PR DESCRIPTION
**Pull request by jblebrun**
Discovered a BackingStore race while debugging an e2e test.

While trying to write a test to verify the BackingStore race fix, I
discovered a `VolatileDriver` init race, as well.

### CODE REVIEW

File: javatests/arcs/core/storage/BackingStoreTest.kt [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import org.junit.Test
+
+class BackingStoreTest {
+
+
```
Comment (jasonwyatt) [inline javatests/arcs/core/storage/BackingStoreTest.kt:22]:
```suggestion
```

File: javatests/arcs/core/storage/BackingStoreTest.kt [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import arcs.core.storage.api.DriverAndKeyConfigurator
+import arcs.core.storage.keys.RamDiskStorageKey
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import org.junit.Test
+
+class BackingStoreTest {
```
Comment (jasonwyatt) [inline javatests/arcs/core/storage/BackingStoreTest.kt:20]:
```suggestion
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class BackingStoreTest {
```

 (not 100% sure on the imports here)

File: javatests/arcs/core/storage/BackingStoreTest.kt [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+
+    @Test
+    fun backingStoreNoRace() = runBlocking<Unit>(Dispatchers.IO) {
+
```
Comment (jasonwyatt) [inline javatests/arcs/core/storage/BackingStoreTest.kt:25]:
```suggestion
```

File: javatests/arcs/core/storage/BackingStoreTest.kt [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        // Attempt to trigger a child store setup race
+        coroutineScope {
+            launch { backingStore.getLocalData("a") }
+            launch { backingStore.onProxyMessage(ProxyMessage.ModelUpdate(data, 1), "a") }
+            launch { backingStore.getLocalData("a") }
+            launch { backingStore.onProxyMessage(ProxyMessage.ModelUpdate(data, 1), "a") }
+            launch { backingStore.getLocalData("a") }
+            launch { backingStore.onProxyMessage(ProxyMessage.ModelUpdate(data, 1), "a") }
+
```
Comment (jasonwyatt) [inline javatests/arcs/core/storage/BackingStoreTest.kt:72]:
```suggestion
```

File: java/arcs/core/storage/BackingStore.kt [hunk 5]
```diff
@@ -88,7 +84,12 @@ class BackingStore<Data : CrdtData, Op : CrdtOperation, T>(
 
         // Return a new Record and add it to our local stores, keyed by muxId.
         return StoreRecord(id, store)
-            .also { record -> storeMutex.withLock { stores[referenceId] = record } }
+    }
+
+    suspend fun store(id: String) = storeMutex.withLock {
```
[the remark this answers is not in the document]
Comment (jblebrun) [reply java/arcs/core/storage/BackingStore.kt:89]: Preferably, but some tests are currently using it. Will mark w/ annotation.

**Pull request merged**